torch/fx/experimental/symbolic_shapes.py

```diff
@@ -3624,10 +3624,12 @@ def issue_guard(guard: ShapeGuard) -> None:
                 raise
 
         # First, issue all the non-trivial guards.
+        previous_guards = ()
         for guard in self.guards:
-            if self._maybe_evaluate_static(guard.expr) is not None:
+            if self._maybe_evaluate_static(guard.expr, axioms=previous_guards) is not None:
                 continue
             issue_guard(guard)
+            previous_guards += (guard.expr,)
```

Comment:
Why not a list and append to it?

Reply:
because we are caching on the args of the function, so we need it to be hashable

Reply:
This seems suspicious. As axioms can be very large, you are essentially forcing an extremely large key in the cache entry, and you're also guaranteed to miss because the axioms list is always change on each iteration.

It seems better to just pass it empty axioms for all input calls, and rely on on-the-fly simplification when we were accumulating...

Reply:
Yeah, I guess that this pass is not necessary now, right?

Reply:
Well it is for now, as otherwise we would be emitting guards of the form `size()[0] <= 10`  and then `2 < = size()[0] <= 10`. We could remove the latter one and just keep the former + the lower bound, but I figured it'd be better to just keep the more compact form `2 <= size()[0] <= 10` for now, so leaving as-is + left a comment.

Reply:
I think you gotta do a benchmark run then, this is potentially quadratic and that makes me very nervous

Reply:
Also, value range refinements already nuke their corresponding guards, so you should never be seeing both of these anyways.

Reply:
no, I removed passing the info iteratively in favour of not passing any axioms. This will not be as strong as it could be, but that's fine.

Regarding "nuking their guards", I guess that that was happening in this code. The life of a check is:
1. Try to prove that `size < 10` and fail
2. Refine the value range to be `sizes \in [2, 9]
3. In this code, prove that `size < 10` is not necessary via value ranges
4. Emit the guard `2 <= size <= 9`

If we remove 3, we end up with the extra guard. My point is that we could potentially delete 3 and just emit `2 <= size` if the lowerbound of `size` is 2. If it's anything else, don't emit it, as that means that it follows from the guards.

Reply:
Ah. So, essentially, what I am hoping is that during the online process, if we see a guard that is 100% subsumed by the value range refinement, I hope we do not put the guard into the list at all. I don't remember if this is what we actually do or not, but that's my ideal situation.

Reply:
That will happen, but it will not happen if the guard itself is the one that triggers the value range refinement. We could make it so that this is the case tho, in which case we could remove the simplifications from this for-loop an simply happily emit all the guards.